lib/ain-evm/src/blocktemplate.rs

```diff
@@ -0,0 +1,376 @@
+    ///
+    /// # Errors
+    ///
+    /// Returns `BlockTemplateError::NoSuchTemplate` if no block template is associated with the given template ID.
+    ///
+    /// # Safety
+    ///
+    /// Result cannot be used safety unless cs_main lock is taken on C++ side
+    /// across all usages. Note: To be replaced with a proper lock flow later.
+    ///
+    pub unsafe fn get(&self, template_id: u64) -> Result<Arc<BlockTemplate>> {
+        Ok(Arc::clone(
+            self.templates
+                .read()
+                .get(&template_id)
+                .ok_or(BlockTemplateError::NoSuchTemplate)?,
+        ))
+    }
+
+    /// Attempts to add a new transaction to the `BlockTemplate` associated with the provided template ID.
+    /// Nonces for each account must be in strictly increasing order. This means that if the last added transaction
+    /// for an account has nonce 3, the next one should have nonce 4. If a `SignedTx` with a nonce that is not one
+    /// more than the previous nonce is added, an error is returned. This helps to ensure the integrity of the block
+    /// template and enforce correct nonce usage.
+    ///
+    /// # Errors
+    ///
+    /// Returns `BlockTemplateError::NoSuchTemplate` if no block template is associated with the given template ID.
+    ///
+    /// # Safety
+    ///
+    /// Result cannot be used safety unless `cs_main` lock is taken on C++ side
+    /// across all usages. Note: To be replaced with a proper lock flow later.
+    ///
+    pub unsafe fn push_in(&self, template_id: u64, tx_update: TxState, hash: XHash) -> Result<()> {
+        self.with_block_template(template_id, |template| template.add_tx(tx_update, hash))
+            .and_then(|res| res)
+    }
+
+    /// Removes all transactions in the block template whose sender matches the provided sender address.
+    /// # Errors
+    ///
+    /// Returns `BlockTemplateError::NoSuchTemplate` if no template is associated with the given template ID.
+    ///
+    /// # Safety
+    ///
+    /// Result cannot be used safety unless `cs_main` lock is taken on C++ side
+    /// across all usages. Note: To be replaced with a proper lock flow later.
+    ///
+    pub unsafe fn remove_txs_above_hash_in(
+        &self,
+        template_id: u64,
+        target_hash: XHash,
+    ) -> Result<Vec<XHash>> {
+        self.with_block_template(template_id, |template| {
+            template.remove_txs_above_hash(target_hash)
+        })
+        .and_then(|res| res)
+    }
+
+    ///
+    /// # Safety
+    ///
+    /// Result cannot be used safety unless cs_main lock is taken on C++ side
+    /// across all usages. Note: To be replaced with a proper lock flow later.
+    ///
+    pub unsafe fn get_txs_cloned_in(&self, template_id: u64) -> Result<Vec<TemplateTxItem>> {
+        self.with_block_template(template_id, BlockTemplate::get_cloned_transactions)
+    }
```

Comment:
Unused, can remove

Reply:
Resolved in commit [ef93279](https://github.com/DeFiCh/ain/pull/2566/commits/ef93279ac70962c9899eb47e0d00c8bbf559153b)